src/forwarder.rs

```
@@ -0,0 +1,387 @@
+        }
+    }
+
+    async fn handle_session_state_task<P>(publisher: P, mut rx_state: Receiver<SessionState>)
+    where
+        P: Publisher,
+    {
+        while let Some(msg) = rx_state.recv().await {
+            let ipath = format!("/{}/status", msg.token);
+            let idata = msg.into();
+
+            if let Err(err) = publisher
+                .send(
+                    "io.edgehog.devicemanager.ForwarderSessionsState",
+                    &ipath,
+                    idata,
+                )
+                .await
+            {
+                error!("publisher send error, {err}");
+            }
+        }
+    }
+
+    /// Start a device forwarder instance.
+    pub fn handle_sessions(&mut self, astarte_event: AstarteDeviceDataEvent) {
+        let Some(idata) = Self::retrieve_astarte_data(astarte_event) else {
+            return;
+        };
+
+        // retrieve the Url that the device must use to open a WebSocket connection with a host
+        let cinfo = match retrieve_connection_info(idata) {
+            Ok(cinfo) => cinfo,
+            // error while retrieving the connection information from the Astarte data
+            Err(err) => {
+                error!("{err}");
+                return;
+            }
+        };
+
+        let bridge_url = match Url::try_from(&cinfo) {
+            Ok(url) => url,
+            Err(err) => {
+                error!("invalid url, {err}");
+                return;
+            }
+        };
+
+        // check if the remote terminal task is already running. if not, spawn a new task and add it to the collection
+        let tx_state = self.tx_state.clone();
+        let session_token = cinfo.session_token.clone();
+        self.get_running(cinfo).or_insert_with(||
+            // spawn a new task responsible for handling the remote terminal operations
+            Self::spawn_task(bridge_url, session_token, tx_state));
+    }
+
+    fn retrieve_astarte_data(
```

this function should return an error since the interface that was passed is incorrect

---

Like retrieve connection info